File: IT/2/Programmazione funzionale/Esercizi/Python/esercizio19.py
# file esercizio19.py

from __future__ import print_function
from itertools import groupby
from random import randint, randrange
from sys import version_info

if version_info[0] > 2: from functools import reduce

def mode(li):
	freq = [(key, len(list(grp))) for key, grp in groupby(sorted(li))]
	md = max(freq, key=lambda t : t[1])[1]
	return list(filter(lambda t : t[1] == md, freq))

l = randrange(5, 100+1, 5)
li = [randint(1, 100) for md in range(l)]

print(li)

# valori con relativa frequenza

fv = mode(li)

md = fv[0][1]

if md == 1:
	print("La lista non ha una moda.")
	exit()

# stampa la moda

print("Moda ({} occorrenze): ".format(md), end="")

if len(fv) == 1:
	print(fv[0][0])
else:
	lm = [t[0] for t in fv]
	print(reduce(lambda m1, m2: "{}, {}".format(m1, m2), lm))

File: IT/2/Programmazione funzionale/Esercizi/C++/esercizio19.cpp
/*
file esercizio19.cpp
compilare con: g++ -std=c++11 esercizio19.cpp
*/

#include <iostream>
#include <iterator>
#include <algorithm>
#include <map>
#include <set>
#include <vector>
#include <functional>
#include <random>

using std::cout;
using std::endl;
using std::ostream_iterator;
using std::random_device;
using std::default_random_engine;
using std::uniform_int_distribution;
using std::vector;
using std::map;
using std::make_pair;
using std::pair;
using std::set;
using std::count;
using std::inserter;
using std::max_element;

vector<pair<int, int>> mode(vector<int>& v) {
	map<int, int> freq;
	set<int> s(v.begin(), v.end());

	transform(s.begin(), s.end(), inserter(freq, freq.begin()), [&] (const int i) {
		int c = count(v.begin(), v.end(), i);
		return make_pair(i, c);
	});

	auto max = *max_element(freq.begin(), freq.end(), [](const pair<int, int>& p1, const pair<int, int>& p2) {
		return p1.second < p2.second;
	});

	int md = max.second;

	vector<pair<int, int>> fv;
	copy_if(freq.begin(), freq.end(), back_inserter(fv), [&](const pair<int, int>& p) {
		return p.second == md;
	});

	return fv;
}

int main()
{
	default_random_engine gen(random_device{}());
	uniform_int_distribution<int> rrange(1, 20);

	vector<int> v(rrange(gen)*5);

	uniform_int_distribution<int> dist(1, 100);
	generate(v.begin(), v.end(), bind(dist, ref(gen)));

	cout << "[";
	copy(v.begin(), v.end(), ostream_iterator<int>(cout, ", "));
	cout << "\b\b]" << endl;

	// valori con relativa frequenza

	vector<pair<int, int>> fv = mode(v);

	int md = fv[0].second;

	if (md == 1) {
		cout << "Il vettore non ha una moda." << endl;
		return 0;
	}

	// stampa la moda

	cout << "Moda (" << md << " occorrenze): ";

	for_each(fv.begin(), fv.end(), [](const pair<int, int>& p) {
		cout << p.first << ", ";
	});

	cout << "\b\b " << endl;
}

